app/models/armor.rb

```diff
@@ -59,13 +62,18 @@ def tempering_materials
   private
 
   def set_canonical_armor
-    return unless canonical_models.count == 1
+    canonicals = canonical_models
```

Comment:
The `#canonical_models` method will always do database calls, which can be expensive. For that reason, we should only call it once here. I've made the same change in the other classes.